### ISSUE TITLE
Closer distance between productionssites and warehouses (by AI)

### ISSUE THREAD

**Issue by MarkMcWire** [widelands/widelands#4930]
I noticed on large maps that the AI often placed production buildings far away in their territory, far away from warehouses. Mostly at the end of a dead end street. This leads to long transport routes and low productivity, which delays the game (with AI versus AI) enormously.

It would be nice if the AI management defined a maximum distance from a production building to the nearest warehouse or port. At least if there is a corresponding building site.

In addition, the AI generally does not build any or only very few warehouses and then usually close to headquarters. It would be nice if there was a minimum distance between the warehouses so that they can be better distributed.

And it is clear to me that the performance of the game drops significantly if there are too many warehouses. I placed warehouses on the map myself using a script and from 5-10 warehouses per player the game shows an enormous drop in the FPS.

**Comment by hessenfarmer** [comment]
This is not a matter of performance drop, but a matter of having a good algorithm to determine this.  Maybe we can try to include some weight on the field that evaluate distance to warehouses. I can't promise anything.

**Comment by MarkMcWire** [comment]
You could do it differently. If a production building is below 25% productivity for a certain period of time, just tear it down. Then the AI has to build it again anyway. Then the buildings that are closer to the warehouse automatically have an advantage.

Exceptions should only be made for mines or when build materials are rare.

**Comment by hessenfarmer** [comment]
That is already implemented to a certain amount. However we do not check every building every time due to performance reasons.     
Difficulty in your suggestion is:   
from distance to warehouise alone a productionsite doesn't drop below 25% if it can be supplied normally.    
tearing down and rebuilding takes a lot of time and material.     
exceptions that are valid for a majority of maps and tribes are hard to define

**Comment by MarkMcWire** [comment]
Okay, then a quick check would be necessary when building a building. The code would have to iterate over all warehouses (including ports) and determine the distance. Then simply impose penalty points for the construction site based on this distance.

Abort the construction from a certain value and search again.

**Comment by hessenfarmer** [comment]
that is not how our AI works. we first iterate over all possible buildable buildings and determine their necessity.   Then we select an amount of fields to check dependent on gametime and size of the economy. for this fields we iterate over all needed (or forced or allowed) buildings and determine the score of each of these buildings on each of these fields. in the end the highest score of a building field combination gets built.      
Every suggestion on the AI needs to be based on this mechanism as changing these basics would not be doable for a single person or at most 2 people who are dealing with the AI currently.

**Comment by MarkMcWire** [comment]
> Then we select an amount of fields

Then it would be possible to limit the number of fields available. Simply remove the fields whose coordinates are furthest away from a warehouse before the other check. In this way, unfavorable fields do not even appear in the list of buildable fields.

You can always take out the for instance 20% of the most distant fields. So if 200 fields are checked, the 40 furthest away are removed, the other 160 are then checked for construction.

The only disadvantage of this method is if there are only suitable places for mines, fishers, etc., which are very far from the nearest warehouse. Then the AI cannot built a mine or fishers house until a warehouse has been set up closer to these fields or all nearby fields have been built on, so that these fields are also taken into account in later tests.

**Comment by hessenfarmer** [comment]
the problem in excluding fields is exactly some productionsites with special needs. So exclusions are made carefully currently. AFAIK only fields near port spaces are excluded for some period if no port is built yet.   
Futhermore distance of fields to warehouses is not calculated yet and might be very performance hungry due to distance is not the real measure but walkable distance is which needs some kind of A-Star algorithm.

**Comment by MarkMcWire** [comment]
I would just use "straight line" as the distance. So the x, y coordinates of the warehouse and the x, y coordinates of the fields. This can be calculated relatively quickly with Pythagoras. I also use this in my "connect road" routine.

**Comment by Noordfrees** [comment]
We already have a C++ function that does the hexagonal-grid maths more efficiently (and much more accurately) than Pythagoras. But to avoid any misunderstandings, before using words like "simply", "just", "quickly" in every post I can recommend [this read](https://www.theregister.com/2012/11/09/bofh_2012_episode_11/) to illustrate how they might (with a touch of programmers' humour) be interpreted 😉

**Comment by tibor95** [comment]
we already have distance any_flag <-> nearest_warehouse calculated, but it is not used for building productionsites at all. I will have a look and see.

**Comment by tibor95** [comment]
Probably easier thing that can be done right now is for every productionsite (trainingsite too?) &buildable_field we can decrease calculated score by distance to nearest warehouse. Or some multiply, maybe managed by genetic algorithm

**Comment by hessenfarmer** [comment]
It might be doable to store distance from nearest warehouse for every field when constructing or destructing / loosing a warehouse going over all fields centrically from the warehouse. We can't do this in calculating buildable fields for performance reasons I think.

**Comment by tibor95** [comment]
But this is already done - for flags - look for flag_warehouse_distance object, It is one of AI jobs.

Do not forget changes in the roads affect these numbers too...

**Comment by tibor95** [comment]
I was thinking to apply this only to productionsites with inputs. What do you think?

**Comment by tibor95** [comment]
Well, I am wrong. We indeed need it for buildable fields, not flags...

**Comment by tibor95** [comment]
Perhaps we can walk from a buildable field and use this info from better accessible flag. I will look at the code

**Comment by hessenfarmer** [comment]
> I was thinking to apply this only to productionsites with inputs. What do you think?

Makes sense. however there might be exceptions which will need evaluation if we are going to implement it. (e.g PS with inputs and space consuming like frisian claypit, or PS with input and strong dependency on supporters like frisian aqua farm or charcoal burner)

**Comment by hessenfarmer** [comment]
> Perhaps we can walk from a buildable field and use this info from better accessible flag. I will look at the code

Well we are walking a map region in most cases however I don't know whether we walk it in circles. If yes we could catch the first flag and take this info as a pseudo measurement which isn't very exact and the error gets greater the more distance from our filed the flag has, as we do not know whether we are nearer to a warehouse as  this flag or further away.

**Comment by tibor95** [comment]
Looking into update_buildable_field(), I see two options:
1. use one of current map.find_immovables(...), this does not check reachability but should get us all flags so we can pick best (lowest) number
2. create new code with map.find_reachable_fields(...) - and this would consider reachability but mean extra CPU use

we can consider option 1 as better alternative of just measuring an air distance to nearest warehouse.

**Comment by hessenfarmer** [comment]
> Looking into update_buildable_field(), I see two options:
> 
>     1. use one of current map.find_immovables(...), this does not check reachability but should get us all flags so we can pick best (lowest) number

a. I wan't to get rid of this helper function find_immovables, because it is very unfriendly for Multithreading (see the AI Multithreading branch)
b. we should not pick lowest as we have a circular region, so average would be better imho. but I agree to the principle of collecting the flags.    

> 
>     2. create new code with map.find_reachable_fields(...) - and this would consider reachability but mean extra CPU use    
    
well reachability isn't what is needed as we need distance, I believe. Calculatin distance would mean find a warehouse first which might be more expensive then calculating the distance 

> 
> 
> we can consider option 1 as better alternative of just measuring an air distance to nearest warehouse.

yes I agree

**Comment by tibor95** [comment]
re using reachability - this would confirm that such flag is connectable by new possible road.

If you want to get rid of find_immovables() is it good to use it? And I like the idea of calculating the average from all flags.

I will look at your branch to check what alternatives are you using

**Comment by tibor95** [comment]
Do you think I can copy this block of code from your branch: https://github.com/widelands/widelands/blob/8e6d8ee835db446a018516276eac2bbf683be2c7/src/ai/defaultai.cc#L1741

And update it as we discussed above?

**Comment by hessenfarmer** [comment]
Well,
simply copying might not be good for 2 reasons:
1. I had to make all variables non static for multithreading.
2. I splitted Military and economic aspects to enable shorter Mutex locks. But this was before changing to directly walk over the mapfields rather then using find_immovables. 

so my preferred solution would be to apply the concept of iterating over the map directly to the current structure (only one funtion instead of 2 separated functions) and try to improve performance by first iterating over the inner region with radius (kProductionArea + 2) for productionsite AND military aspects. and afterwards only iterate over the outer area (map hollow region with inner radius kProductionArea + 2 and outer radius actual_enemy_check_area) for military aspects only.
this would save us iterating over the inner fields twice like it is done now.

**Comment by tibor95** [comment]
Reasonable. I will go that way

**Comment by tibor95** [comment]
I need your advice and approval.
I have an idea to take out chucks of code from update_buildable_field to separate methods, see partial implementation here:
https://github.com/widelands/widelands/blob/6644f3159133f833307737834a6cf3f531ae7e62/src/ai/defaultai.cc#L1857

look for handle_own_ps() method. The idea is to create another 1-3 such methods, and this would help us with unifying those two radius checks, improve code readability and so.

Do you agree to go this way?

**Comment by hessenfarmer** [comment]
the concept looks good to go.

However 2 things I found in your partial implementation:
1. at this stage in the code we are only in interested in Buildings so you should check first if it is a building and only continue if we have a building ( see  https://github.com/widelands/widelands/blob/8e6d8ee835db446a018516276eac2bbf683be2c7/src/ai/defaultai.cc#L1756)
2. you don't need the unique_coords mechanism as in walkin a mapregion the ccord is always unique. But you need the serial of the building to be unique as in 
https://github.com/widelands/widelands/blob/8e6d8ee835db446a018516276eac2bbf683be2c7/src/ai/defaultai.cc#L1760
because otherwise big buildings would be counted multiple times as they occupy mor then one field.

**Comment by tibor95** [comment]
re 1.: - Fair comment, but for the future we are interested also in flags here, so test should be 'flag or building'.

re 2.: - I completely forgot about big buildings :(

I was also afraid about pointers and references here, and also bit struggling to make it compile.

Small question, is it possible that building will disappear while update_buildable_fields() is being executed? I presume NOT.

**Comment by hessenfarmer** [comment]
in multithreading it is possible and it is probable as well, as the economy code is executed in parallel there. This was the reason for me to not use find_immovables anymore in the multithreading branch. So to prepare for making the AI multithreadable in the future we should aim at keeping long living pointers that can be changed outside the AI to an absolute minimum.

**Comment by tibor95** [comment]
I progressed a bit, in fact the refactoring is more-less done. However I was not able to create that hollow area, can you please have a look?:
https://github.com/widelands/widelands/blob/079a06a739caa2ff9f91593ff8fb5db273da158a/src/ai/defaultai.cc#L1903
That first commented line is my attempt. Also if you have time you can have a look on other stuff around that I changed.
Thanks

**Comment by hessenfarmer** [comment]
for the hollow area see
https://github.com/widelands/widelands/blob/7bb011c1efb4b8a66b0ab567b64f3095a814400f/src/ai/defaultai.cc#L1233

**Comment by hessenfarmer** [comment]
will have alook into the rest later on.
maybe you might open a draft PR to make commenting /suggeting easier

**Comment by tibor95** [comment]
What is difference between HollowArea and MapHollowRegion? Do I need the later, or the first is usable?

**Comment by tibor95** [comment]
https://github.com/widelands/widelands/pull/5216 - we can continue there with discussion

**Comment by hessenfarmer** [comment]
> What is difference between HollowArea and MapHollowRegion? Do I need the later, or the first is usable?

HollowArea defines just a donut style area shape MapHollowRegion expects such a shape to provide each Node in this area to be examined.  
So you need both of them: 
1. define the Area
2. pass this Area to the MapHollowRegion

**Comment by tibor95** [comment]
https://github.com/widelands/widelands/blob/3d33a62278414b0136c23e8799721723346f3788/src/ai/defaultai.cc#L1903

I tried, but compilation fails on:

```
[build] /usr/bin/ld: ../libai.a(defaultai.cc.o): in function `AI::DefaultAI::update_buildable_field(AI::BuildableField&)':
[build] /var/widelands2/git/trunk/src/ai/defaultai.cc:1904: undefined reference to `Widelands::MapHollowRegion<Widelands::Area<Widelands::FCoords, unsigned short> >::MapHollowRegion(Widelands::Map const&, Widelands::HollowArea<Widelands::Area<Widelands::FCoords, unsigned short> > const&)'
[build] /usr/bin/ld: /var/widelands2/git/trunk/src/ai/defaultai.cc:1941: undefined reference to `Widelands::MapHollowRegion<Widelands::Area<Widelands::FCoords, unsigned short> >::advance(Widelands::Map const&)'
[build] collect2: error: ld returned 1 exit status
```

Can you advice?
